### DISCUSSION TITLE
Is iperf suitable for testing in a 400G environment?

### DISCUSSION THREAD

**Discussion by woojuazoomma** [esnet/iperf#1594] [Q&A]
"When testing with 4 hosts using a 2-port 200G NIC - mcx653106A-HDAT-SP, and running the command 'iperf -c 10.X.X.X -T s1 -P 8 -l 32768 -w 35M -t 30s,' the combined result shows 200G. However, we desire a total of 400G from the client. Are there any specific command options or considerations during the test that need to be verified? Is it possible that iperf testing has a limit of 200G?"

**Comment by bmah888** [comment]
I understand you're expecting to get 400Gbps of throughput through a 2x200Gbps NIC, but you only got 200Gbps.

There aren't a lot of details in your post, but it seems likely that all of the iperf3 traffic on this host are going through one of the 200Gbps ports, and the other port is idle. To get both of the ports to be used, you probably need to do some some of link aggregation ("lag") to get the two ports to act as a single 400Gbps port. Even then, it might not do exactly what you want, due to the way that traffic is split between the different components of a lag.

Actually as I'm looking at the specifications for the card (https://store.nvidia.com/en-us/networking/store/product/mcx653106a-hdat-sp/nvidia-connectx-6-vpi-adapter-card-hdr-200gbe/), they mention a maximum bandwidth of 200Gbps, so it might not be possible to ever get more throughput than that.

As far as we're aware, iperf3 doesn't have any intrinsic limits.

**Reply by woojuazoomma** [reply]
"Thank you for your response. We need to confirm the ability to achieve a throughput of 400G from four servers ('A') with 100G NICs each to a single server ('B') with a 200G 2-port X2 (200NIC, 2 ports) configuration, using IPERF3. Despite consulting documentation and conducting tests, we are unable to achieve more than 20G. On the client side (A, four servers), we are using the following command: 'iperf3 -c 10.x.x.xxx -T s1 -l 32768 -t 5 -p 520.' On the server side (B), the command used is 'iperf3 -s.' Could you please confirm if the testing commands we are using are correct? Additionally, could you provide the appropriate commands or reference documentation for our specific scenario?"
***The translation through ChatGPT may not convey 100% of the intended meaning in English.

**Reply by bmah888** [reply]
Your commands look OK. However there are a few reasons why you might not be able to get your desired results. I don't know which of these apply to your situation, but they might include:

1. It's not possible to get the full capability of the two 200G ports, according to the specifications the card is limited to a total throughput of 200Gbps, not 400Gbps. It's actually very difficult for a host architecture to send or receive 400Gbps due to the limits of the PCIe bus on commonly available x86 servers. Also this NIC is a PCIe Gen 4 card and PCIe Gen 4 has a maximum of 200Gbps.
2. The iperf3 server is only designed to handle a single test at a time. This limitation is done on purpose, to avoid tests interfering with one another. So it's not possible to do tests from 4 clients to the same server at the same time.
3. You didn't give details about how the four 'A' servers are connected to the single 'B' server, both physically (cabling) and logically (IP addressing, etc.). Depending on how these are connected it might not be possible to use both ports at the same time on the card in server 'B'.
4. You didn't say what version of iperf3 you're using, but versions of iperf3 before 3.16 are limited to using a single processor core across all simultaneous tests. The most throughput we have seen is around 70Gbps. Later versions of iperf3 (3.16 or newer) can use multiple cores and have a better chance of saturating the links.

Hope this helps!
